File: app/src/main/java/es/etologic/mahjongscoring2/app/base/BaseActivity.java
/*
 * Copyright © 2018 LaLiga.
 * All rights reserved.
 */

package es.etologic.mahjongscoring2.app.base;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.View;
import android.widget.LinearLayout;

import org.apache.commons.lang3.StringUtils;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import dagger.android.support.DaggerAppCompatActivity;
import es.etologic.mahjongscoring2.R;
import es.etologic.mahjongscoring2.app.model.ShowState;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static es.etologic.mahjongscoring2.app.model.ShowState.SHOW;

@SuppressLint("Registered")
public class BaseActivity extends DaggerAppCompatActivity {

    //FIELDS
    @BindView(R.id.llProgress) LinearLayout llProgress;
    private Unbinder viewUnbinder;

    //LIFECYCLE
    @Override public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);
        viewUnbinder = ButterKnife.bind(this);
    }
    public void goToFragment(Fragment fragment, @IdRes int frameLayoutContainer) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.setCustomAnimations(android.R.animator.fade_in, android.R.animator.fade_out);
        fragmentTransaction.replace(frameLayoutContainer, fragment);
        fragmentTransaction.commit();
    }
    public void goToActivity(Intent intent, int requestCode) {
        startActivity/*ForResult*/(intent/*, requestCode*/);
//        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left);
    }
    @Override protected void onDestroy() {
        if (viewUnbinder != null) {
            viewUnbinder.unbind();
        }
        super.onDestroy();
    }
    protected void showError(Throwable throwable) {
        showError(throwable == null ? getString(R.string.ups_something_wrong) : throwable.getMessage());
    }
    private void showError(String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.AppTheme);
        builder.setMessage(message)
//                .create()
                .show();
    }
    protected void showSnackbar(View view, String message) {
        String text = StringUtils.isBlank(message) ? "" : message;
        Snackbar.make(view, text, Snackbar.LENGTH_LONG)
                .show();
    }

    protected void toggleProgress(ShowState showState) {
        llProgress.setVisibility((showState == SHOW) ? VISIBLE : GONE);
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/base/BaseActivity.kt
/*
 * Copyright © 2018 LaLiga.
 * All rights reserved.
 */

package es.etologic.mahjongscoring2.app.base

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Intent
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.Snackbar
import dagger.android.support.DaggerAppCompatActivity
import es.etologic.mahjongscoring2.R
import es.etologic.mahjongscoring2.app.model.ShowState
import es.etologic.mahjongscoring2.app.model.ShowState.SHOW
import kotlinx.android.synthetic.main.main_activity.*
import org.apache.commons.lang3.StringUtils

@SuppressLint("Registered")
open class BaseActivity : DaggerAppCompatActivity() {
    
    protected fun goToFragment(fragment: Fragment, @IdRes frameLayoutContainer: Int) {
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.setCustomAnimations(android.R.animator.fade_in, android.R.animator.fade_out)
        fragmentTransaction.replace(frameLayoutContainer, fragment)
        fragmentTransaction.commit()
    }
    
    protected fun goToActivity(intent: Intent, requestCode: Int) {
        startActivity/*ForResult*/(intent/*, requestCode*/)
    }
    
    internal fun showError(throwable: Throwable?) {
        if (throwable == null) getString(R.string.ups_something_wrong)
        else throwable.message?.let { showError(it) }
    }
    
    protected fun showError(message: String) {
        val builder = AlertDialog.Builder(this, R.style.AppTheme)
        builder.setMessage(message)
            .show()
    }
    
    internal fun showSnackbar(view: View, message: String) {
        val text = if (StringUtils.isBlank(message)) "" else message
        Snackbar.make(view, text, Snackbar.LENGTH_LONG)
            .show()
    }
    
    protected fun toggleProgress(showState: ShowState) {
        llProgress?.visibility = if (showState === SHOW) VISIBLE else GONE
    }
}
